core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala

```diff
@@ -18,11 +18,21 @@
 package org.apache.spark.scheduler
 
 /**
- * Provides more detail when an executor is being decommissioned.
+ * Message providing more detail when an executor is being decommissioned.
  * @param message Human readable reason for why the decommissioning is happening.
  * @param isHostDecommissioned Whether the host (aka the `node` or `worker` in other places) is
  *                             being decommissioned too. Used to infer if the shuffle data might
  *                             be lost even if the external shuffle service is enabled.
  */
 private[spark]
 case class ExecutorDecommissionInfo(message: String, isHostDecommissioned: Boolean)
+
+/**
+ * State related to decommissioning that is kept by the TaskSchedulerImpl. This state is derived
+ * from the info message above but it is kept distinct to allow the state to evolve independently
+ * from the message.
+ */
+case class ExecutorDecommissionState(message: String,
```

Comment:
why not `(info: ExecutorDecommissionInfo, tsMillis: Long)`?

Reply:
I actually toyed with that and then backed out: While it does save some "redundancy". The decoupling would allow for evolution here because this decommissioning project is so new. I expect the state keeping and the messages to evolve quite rapidly in the beginning and having this flexibility would be helpful. Just a philosophical thought :-P. We can always converge it more later. 

But having said that, I am not opposed to embedding the info message in.

Reply:
I'd even prefer `ExecutorDecommissionInfo(message: String, tsMillis: Long, isHostDecommissioned: Boolean)` to make codes simpler...and I don't understand "messages to evolve quite rapidly"...the `message` parameter is a `String` type, it should support any arbitrary strings. Why it's a problem?

Reply:
I consider `ExecutorDecommissionInfo` a message because it is a part of the contract b/w the AppClientListener and Scheduler. It is also embedded inside the RPC message DecommissionExecutor. So I felt that adding the timestamp field to a "message' is not a good idea because it bloats it un-necessarily with information it does not need. It also raises the question what does that timestamp even mean ? Timestamp of the executor (which can originate this message) or timestamp of the driver ? 

Its not a problem to couple these -- its just a preference to not couple them so that we can change them as time goes by (evolve them). Having a copy is redundant but it is also flexible.

By "messages evolving rapidly", sorry I meant the case class `ExecutorDecommissionInfo` and not the message field inside it. :-P

Reply:
> So I felt that adding the timestamp field to a "message' is not a good idea because it bloats it un-necessarily with information it does not need. It also raises the question what does that timestamp even mean ? Timestamp of the executor (which can originate this message) or timestamp of the driver ?


The timestamp will be used to calculate the `canExceedDeadline` in `TaskSetManager` later? And the timestamp means the approximate time the executor starts to decommission.


> Its not a problem to couple these -- its just a preference to not couple them so that we can change them as time goes by (evolve them). Having a copy is redundant but it is also flexible.

I just can not imagine what kind of evolution would make things to be hard to handle after we couple them. Actually, what worries me more is that `ExecutorDecommissionInfo` might not satisfy the requirement of other clusters(if we'd support them in the future). That could be a reason that we should not couple `ExecutorDecommissionState` with `ExecutorDecommissionInfo`. This also reminds me again that we'd better have an interface for `ExecutorDecommissionInfo` like `ExecutorLossReason`, so we could have a way to work with other cluster managers rather than Standalone only.

Reply:
Decoupling `ExecutorDecommissionInfo` and `ExecutorDecommissionState` sounds reasonable, +1.

Reply:
@Ngone51, Good point about the other cluster managers. I would bet that we need to tweak `ExecutorDecommissionInfo` if we support YARN more fully. And yeah that is another reason to decouple the state keeping from the actual message. We are more free to change the state since it is used for internal book keeping than changing the `ExecutorDecommissionInfo`.

It seems that we are all on the same page and that we are okay with this decoupling of ``ExecutorDecommissionInfo` and `ExecutorDecommissionState` ? Let me know if I misread your comment and that is not yet the case.

Reply:
Yes, decoupling +1